File: ios/Sources/Drengr/Drengr.swift
import Foundation
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

/// Zero-code network analytics for iOS.
///
///     Drengr.start(
///         publishableKey: "drengr_pk_…",                                   // from drengr.dev/pro
///         ingestURL: "https://<ref>.supabase.co/functions/v1/ingest",
///         appPackage: "com.example.app"
///     )
///
/// One call installs PASSIVE capture (method-swizzled observation — see
/// `URLSessionCapture`) of every `URLSession` exchange across all sessions
/// (`.shared`, `.default`, custom configs, Alamofire, async/await). Because it
/// only observes — never re-issues — the app's certificate pinning / auth-challenge
/// delegate runs completely untouched. Redaction is on by default; the sink's own
/// delivery uses a separate session it never observes, so it can't self-capture.
///
/// Coverage: metadata (method/url/status/timing/bytes/error) is exact for every
/// task; response BODIES are captured for completion-handler, async/await, and
/// delegate-based (iOS 15+) tasks.
public enum Drengr {
    public static let version = "0.2.0"

    private static let lock = NSLock()
    private static var sink: IngestSink?
    private static var tracker: SessionTracker?
    private static var enabled = true
    private static var installed = false
    private static var observers: [NSObjectProtocol] = []

    /// Initialize capture + delivery. Call once at launch. Subsequent calls are ignored.
    ///
    /// Capture is PASSIVE (method-swizzled observation) — the app's own certificate
    /// pinning / auth-challenge delegate runs untouched, and every session is covered
    /// automatically (no per-configuration registration needed).
    ///
    /// - Parameters:
    ///   - maxBodyBytes: cap on captured body size (default 64 KiB).
    ///   - startEnabled: false installs paused (consent gate); call `setEnabled`.
    ///   - behavior: default ON — captures screen_view (viewDidAppear swizzle),
    ///     taps with semantic labels, rage/dead taps, and crashes with zero
    ///     per-screen code; pass `false` to opt out of behavior autocapture.
    ///   - captureWhen: optional per-URL predicate (sampling / allow-listing).
    ///   - redactHeaders: extra header names to mask (lowercased), on top of defaults.
    ///   - vault: E2EE PII vault pin; sealing stays off until the org's signed
    ///     config is fetched and verifies against it (fail-closed).
    public static func start(
        publishableKey: String,
        ingestURL: String,
        appPackage: String,
        maxBodyBytes: Int = 64 * 1024,
        startEnabled: Bool = true,
        behavior: Bool = true,
        captureWhen: ((String) -> Bool)? = nil,
        redactHeaders: Set<String> = [],
        extraContext: [String: Any] = [:],
        vault: VaultOptions? = nil
    ) {
        lock.lock(); defer { lock.unlock() }
        if installed { return }
        guard let url = URL(string: ingestURL) else { return }

        // A persisted opt-out always wins over startEnabled: an opted-out install
        // stays paused across restarts until optIn() is called (GDPR).
        enabled = startEnabled && !isOptedOut()

        let t = SessionTracker()
        tracker = t
        var ctx: [String: Any] = [
            "app_package": appPackage,
            "os": "ios",
            "os_version": osVersion(),
            "device_model": deviceModel(),
            "install_id": installID(),
            "session_id": t.id,
            "sdk_version": version,
        ]
        for (k, v) in extraContext { ctx[k] = v }
        let s = IngestSink(url: url, publishableKey: publishableKey, context: ctx)
        sink = s

        let lowerExtra = Set(redactHeaders.map { $0.lowercased() })
        URLSessionCapture.install(config: URLSessionCapture.Config(
            maxBodyBytes: maxBodyBytes,
            redactHeaderNames: lowerExtra,
            onEvent: { s.addNetwork($0); t.touch() },
            isEnabled: { enabledFlag() },
            shouldCapture: captureGate(ingestHost: url.host, captureWhen)
        ))
        installLifecycleObservers()
        installed = true

        if behavior { installBehaviorCapture(sink: s) }

        if let v = vault {
            VaultActivation.configure(sink: s, options: v, ingestURL: ingestURL,
                                      publishableKey: publishableKey, session: s.delivery)
        }
    }

    /// Behavior autocapture (default ON via start): taps + rage/dead taps,
    /// screen_view, and crash — all through the sink's redacted event path.
    /// Also drains crash events a previous run's signal handler persisted.
    private static func installBehaviorCapture(sink s: IngestSink) {
        #if canImport(UIKit) && !os(watchOS)
        installTapCapture()
        ScreenViewCapture.install(config: ScreenViewCapture.Config(
            onEvent: { ev in
                lock.lock(); let s = sink; let t = tracker; lock.unlock()
                s?.addBehavior(ev); t?.touch()
            },
            isEnabled: { enabledFlag() }
        ))
        #endif
        // The crash closure captures the sink directly: at crash time it must
        // never contend for Drengr's lock (a wedged lock would hang the crash).
        CrashCapture.install(config: CrashCapture.Config(
            onCrash: { ev in
                s.addCrashSync(ev)
                let sem = DispatchSemaphore(value: 0)
                s.forceFlush { sem.signal() }
                _ = sem.wait(timeout: .now() + 2)   // bounded: never turn a crash into a hang
            }
        ))
        CrashCapture.setEnabled(enabled)   // start() holds the lock — read the flag directly
        CrashCapture.drainPending(into: s)
    }

    private static func installTapCapture() {
        #if canImport(UIKit) && !os(watchOS)
        TapCapture.install(config: TapCapture.Config(
            onEvent: { e in
                lock.lock(); let s = sink; let t = tracker; lock.unlock()
                s?.addTap(e); t?.touch()
            },
            onBehavior: { ev in
                lock.lock(); let s = sink; let t = tracker; lock.unlock()
                s?.addBehavior(ev); t?.touch()
            },
            isEnabled: { enabledFlag() }
        ))
        #endif
    }

    /// Deprecated: tap capture is ON BY DEFAULT in `start` (behavior: true).
    /// Kept for source compatibility; still installs when start(behavior: false)
    /// was used.
    @available(*, deprecated, message: "Tap capture is on by default in start(); this call is no longer needed.")
    public static func experimentalSwiftUITapCapture() {
        installTapCapture()
    }

    /// Force-send all queued events now; `completion` fires when the attempt finishes.
    public static func flush(completion: (() -> Void)? = nil) {
        lock.lock(); let s = sink; lock.unlock()
        guard let s = s else { completion?(); return }
        s.forceFlush(completion)
    }

    // --- session lifecycle (foreground rotation check, background touch + flush) ---

    private static func installLifecycleObservers() {
        let nc = NotificationCenter.default
        #if canImport(UIKit)
        observers.append(nc.addObserver(forName: UIApplication.willEnterForegroundNotification,
                                        object: nil, queue: nil) { _ in onForeground() })
        observers.append(nc.addObserver(forName: UIApplication.didEnterBackgroundNotification,
                                        object: nil, queue: nil) { _ in onBackground() })
        #elseif canImport(AppKit)
        observers.append(nc.addObserver(forName: NSApplication.didBecomeActiveNotification,
                                        object: nil, queue: nil) { _ in onForeground() })
        observers.append(nc.addObserver(forName: NSApplication.didResignActiveNotification,
                                        object: nil, queue: nil) { _ in onBackground() })
        #endif
    }

    static func onForeground() {
        lock.lock(); let t = tracker; let s = sink; let en = enabled; lock.unlock()
        guard let t = t else { return }
        if t.isStale() {
            let newId = t.rotate()
            s?.rotateSession(to: newId)   // old queue flushes under the old id first
        } else {
            t.touch()
        }
        // Foreground marker so the backend can subtract backgrounded time from
        // screen dwell (accurate dwell is the make-or-break signal). Mirrors
        // Android onActivityStarted; rides under the new session id after rotation.
        if en { s?.addAppForeground(ScreenState.screen) }
    }

    static func onBackground() {
        lock.lock(); let t = tracker; let s = sink; let en = enabled; lock.unlock()
        t?.touch()
        // Emit BEFORE the flush so the background marker rides the same batch
        // (both hop the sink's serial queue in FIFO order). Mirrors Android
        // onActivityStopped: app_background, then flush.
        if en { s?.addAppBackground(ScreenState.screen) }
        s?.forceFlush()
    }

    /// Never capture requests to our own ingest host (self-capture loop guard);
    /// then apply the app's optional predicate.
    static func captureGate(ingestHost: String?, _ user: ((String) -> Bool)?) -> (String) -> Bool {
        let host = ingestHost?.lowercased()
        return { urlString in
            if let h = host, URL(string: urlString)?.host?.lowercased() == h { return false }
            return user?(urlString) ?? true
        }
    }

    /// Deprecated no-op: capture is now global (swizzled), so custom sessions are
    /// covered automatically. Kept for source compatibility.
    @available(*, deprecated, message: "No longer needed — all sessions are captured automatically.")
    public static func register(on configuration: URLSessionConfiguration) {}

    /// Pause/resume capture (consent gate). Delivery of buffered events continues.
    public static func setEnabled(_ value: Bool) {
        lock.lock(); enabled = value; lock.unlock()
        CrashCapture.setEnabled(value)
    }

    /// Persistently opt this install OUT of capture (GDPR). Unlike setEnabled(false),
    /// this survives app restarts — start() reads it and stays paused next launch.
    public static func optOut() {
        UserDefaults.standard.set(true, forKey: optOutKey)
        setEnabled(false)
    }

    /// Reverse optOut(): clear the persisted flag and resume capture.
    public static func optIn() {
        UserDefaults.standard.removeObject(forKey: optOutKey)
        setEnabled(true)
    }

    /// Sets external_id — your own stable, non-PII user id (not an email) — on
    /// the session and all events hereafter; emits one identify event. `traits`
    /// are redacted before delivery. Fail-open: no-op if `start` hasn't run yet
    /// or `externalId` is empty.
    public static func identify(_ externalId: String, traits: [String: Any] = [:]) {
        lock.lock(); let s = sink; let t = tracker; lock.unlock()
        s?.identify(externalId, traits: traits)
        t?.touch()
    }

    /// Tags the session with an experiment variant, attached to all events
    /// hereafter as `experiments`. A nil/empty `variant` clears `key`.
    public static func setExperiment(_ key: String, variant: String?) {
        lock.lock(); let s = sink; lock.unlock()
        s?.setExperiment(key, variant: variant)
    }

    private static func enabledFlag() -> Bool {
        lock.lock(); defer { lock.unlock() }; return enabled
    }

    // --- persisted opt-out (mirrors install id) ---
    private static let optOutKey = "dev.drengr.opt_out"
    private static func isOptedOut() -> Bool { UserDefaults.standard.bool(forKey: optOutKey) }

    // --- install id (persisted in UserDefaults) ---
    private static let installKey = "dev.drengr.install_id"
    private static func installID() -> String {
        let d = UserDefaults.standard
        if let existing = d.string(forKey: installKey) { return existing }
        let id = UUID().uuidString
        d.set(id, forKey: installKey)
        return id
    }

    private static func osVersion() -> String {
        let v = ProcessInfo.processInfo.operatingSystemVersion
        return "\(v.majorVersion).\(v.minorVersion).\(v.patchVersion)"
    }

    private static func deviceModel() -> String {
        var size = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        var machine = [CChar](repeating: 0, count: size)
        sysctlbyname("hw.machine", &machine, &size, nil, 0)
        return String(cString: machine)
    }
}

File: android/drengr/src/main/kotlin/dev/drengr/sdk/Drengr.kt
package dev.drengr.sdk

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import okhttp3.OkHttpClient
import java.util.UUID
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Zero-code network analytics for Android.
 *
 *   val client = OkHttpClient.Builder()
 *       .addInterceptor(
 *           Drengr.start(
 *               context = applicationContext,
 *               publishableKey = "drengr_pk_…",
 *               ingestUrl = "https://<ref>.supabase.co/functions/v1/ingest",
 *               appPackage = "com.example.app",
 *           )
 *       )
 *       .build()
 *
 * `start` returns the [DrengrInterceptor] to add to your OkHttp client(s) — the
 * one integration point. Redaction is on by default; the sink's own delivery is
 * a separate client with no interceptor, so it can never capture itself.
 */
object Drengr {
    private val enabled = AtomicBoolean(true)
    private var sink: IngestSink? = null
    private var session: SessionManager? = null
    // Application context (process-lived singleton — safe to hold statically, no leak)
    // so optOut()/optIn() can persist the choice without a Context argument.
    private var appContext: Context? = null
    private val lifecycleRegistered = AtomicBoolean(false)
    @Volatile private var redactExtra: Set<String> = emptySet()
    @Volatile private var maxBody: Long = 64 * 1024
    private const val PREFS = "drengr_sdk"
    private const val INSTALL_KEY = "install_id"
    private const val OPTOUT_KEY = "opt_out"

    const val VERSION = "0.2.0"

    /**
     * Initialize capture + delivery and return the interceptor to install on your
     * OkHttpClient. Call once at startup. Safe to call again (returns a fresh
     * interceptor bound to the same sink).
     *
     * @param maxBodyBytes cap on captured body size (default 64 KiB).
     * @param startEnabled false installs paused (consent gate); call [setEnabled].
     * @param captureWhen optional per-URL predicate (sampling / allow-listing).
     * @param redactHeaders extra header names to mask (lowercased), on top of defaults.
     */
    @JvmStatic
    @JvmOverloads
    fun start(
        context: Context,
        publishableKey: String,
        ingestUrl: String,
        appPackage: String,
        maxBodyBytes: Long = 64 * 1024,
        startEnabled: Boolean = true,
        captureWhen: ((String) -> Boolean)? = null,
        redactHeaders: Set<String> = emptySet(),
        extraContext: Map<String, Any?> = emptyMap(),
    ): DrengrInterceptor {
        val appCtx = context.applicationContext
        appContext = appCtx
        // A persisted opt-out always wins over startEnabled — an opted-out install
        // stays paused across restarts until optIn() (GDPR).
        enabled.set(startEnabled && !isOptedOut(appCtx))
        if (sink == null) {
            val sm = try {
                SessionManager(appCtx.getSharedPreferences(PREFS, Context.MODE_PRIVATE))
            } catch (_: Throwable) { null }
            session = sm
            val ctx = HashMap<String, Any?>().apply {
                put("app_package", appPackage)
                put("os", "android")
                put("os_version", android.os.Build.VERSION.RELEASE ?: "")
                put("device_model", android.os.Build.MODEL ?: "")
                put("install_id", installId(appCtx))
                put("sdk_version", VERSION)
                putAll(extraContext)
            }
            sink = IngestSink(
                appCtx.filesDir, ingestUrl, publishableKey, ctx,
                sessionId0 = sm?.sessionId ?: "s-${System.currentTimeMillis()}",
            )
            registerLifecycle(appCtx)
        }
        val s = sink!!
        redactExtra = redactHeaders.map { it.lowercase() }.toSet()
        maxBody = maxBodyBytes
        return DrengrInterceptor(
            maxBodyBytes = maxBodyBytes,
            redactHeaderNames = redactExtra,
            onEvent = { session?.touch(); s.addNetwork(it) },
            enabled = { enabled.get() },
            shouldCapture = { url -> captureWhen?.invoke(url) ?: true },
        )
    }

    /** Force-send buffered events now. [onComplete] runs after the attempt. */
    @JvmStatic
    @JvmOverloads
    fun flush(onComplete: Runnable? = null) {
        val s = sink
        if (s == null) {
            try { onComplete?.run() } catch (_: Throwable) {}
            return
        }
        s.flushNow(onComplete)
    }

    /**
     * Opt-in HttpURLConnection/HttpsURLConnection capture via a process-global
     * delegating URLStreamHandlerFactory (set-once per process). Call after
     * [start]. Returns false — with one logged warning, never a crash — when
     * another factory already owns the process. Captured exchanges go through
     * the same redaction + delivery path as the OkHttp interceptor.
     */
    @JvmStatic
    fun installUrlConnectionCapture(): Boolean {
        val s = sink ?: return false
        return UrlConnectionCapture.install(
            onEvent = { session?.touch(); s.addNetwork(it) },
            enabled = { enabled.get() },
            redactExtra = redactExtra,
            maxBodyBytes = maxBody,
        )
    }

    /**
     * Cronet adapter (opt-in): returns a `RequestFinishedInfo.Listener` to pass to
     * `CronetEngine.Builder.addRequestFinishedListener`, or null when Cronet is not
     * on the classpath (clean no-op — the SDK never hard-requires Cronet). Typed
     * `Any?` so this method resolves without Cronet present.
     */
    @JvmStatic
    fun cronetListener(executor: Executor): Any? = try {
        Class.forName("org.chromium.net.RequestFinishedInfo")
        DrengrCronetListener(executor) { e ->
            if (enabled.get()) {
                session?.touch()
                sink?.addNetwork(e)
            }
        }
    } catch (_: Throwable) {
        null
    }

    /**
     * EXPERIMENTAL — Compose semantic tap capture. Wraps the activity's window
     * callback to observe taps and resolve the tapped Compose node's semantics
     * (contentDescription/text/role) into labeled tap events. Returns false —
     * never a crash — when compose-ui is absent (clean no-op, compose is never
     * hard-required), [start] hasn't run, or the window isn't ready. Fail-open:
     * touch delivery to the app is never affected.
     */
    @JvmStatic
    fun experimentalComposeTapCapture(activity: Activity): Boolean {
        return try {
            Class.forName("androidx.compose.ui.semantics.SemanticsOwner")
            val s = sink ?: return false
            ComposeTapCapture.install(activity) { e ->
                if (enabled.get()) {
                    session?.touch()
                    s.addTap(e)
                }
            }
        } catch (_: Throwable) {
            false
        }
    }

    /** EXPERIMENTAL — restore the window callback wrapped by [experimentalComposeTapCapture]. */
    @JvmStatic
    fun experimentalComposeTapCaptureStop(activity: Activity): Boolean {
        return try {
            Class.forName("androidx.compose.ui.semantics.SemanticsOwner")
            ComposeTapCapture.uninstall(activity)
        } catch (_: Throwable) {
            false
        }
    }

    private fun registerLifecycle(ctx: Context) {
        if (!lifecycleRegistered.compareAndSet(false, true)) return
        try {
            (ctx as? Application)?.registerActivityLifecycleCallbacks(lifecycle)
        } catch (_: Throwable) {}
    }

    // Foreground = zero→nonzero started activities (rotate stale session);
    // background = nonzero→zero (persist activity + auto-flush).
    private val lifecycle = object : Application.ActivityLifecycleCallbacks {
        private var started = 0

        override fun onActivityStarted(activity: Activity) {
            if (started++ != 0) return
            try {
                val sm = session ?: return
                if (sm.rotateIfStale()) sink?.rotateSession(sm.sessionId)
            } catch (_: Throwable) {}
        }

        override fun onActivityStopped(activity: Activity) {
            if (started > 0) started--
            if (started != 0) return
            try { session?.touch(force = true) } catch (_: Throwable) {}
            try { sink?.flushNow() } catch (_: Throwable) {}
        }

        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
        override fun onActivityResumed(activity: Activity) {}
        override fun onActivityPaused(activity: Activity) {}
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) {}
    }

    /** Pause/resume capture (consent gate). Delivery of buffered events continues. */
    @JvmStatic
    fun setEnabled(value: Boolean) {
        enabled.set(value)
    }

    /**
     * Persistently opt this install OUT of capture (GDPR). Unlike setEnabled(false),
     * this survives app restarts — start() reads it and stays paused next launch.
     * No-op if start() hasn't run yet (no application context available).
     */
    @JvmStatic
    fun optOut() {
        appContext?.let { optOutPref(it, true) }
        enabled.set(false)
    }

    /** Reverse optOut(): clear the persisted flag and resume capture. */
    @JvmStatic
    fun optIn() {
        appContext?.let { optOutPref(it, false) }
        enabled.set(true)
    }

    /**
     * Sets external_id — your own stable, non-PII user id (not an email) — on the
     * session and all events hereafter; emits one identify event. [traits] are
     * redacted before delivery. Fail-open: no-op if [start] hasn't run yet or
     * [externalId] is empty.
     */
    @JvmStatic
    @JvmOverloads
    fun identify(externalId: String, traits: Map<String, Any?> = emptyMap()) {
        sink?.identify(externalId, traits)
    }

    /**
     * Tags the session with an experiment variant, attached to all events
     * hereafter as `experiments`. A null/empty [variant] clears [key].
     */
    @JvmStatic
    fun setExperiment(key: String, variant: String?) {
        sink?.setExperiment(key, variant)
    }

    private fun isOptedOut(ctx: Context): Boolean =
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getBoolean(OPTOUT_KEY, false)

    private fun optOutPref(ctx: Context, value: Boolean) {
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().putBoolean(OPTOUT_KEY, value).apply()
    }

    private fun installId(ctx: Context): String {
        val prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        prefs.getString(INSTALL_KEY, null)?.let { return it }
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(INSTALL_KEY, id).apply()
        return id
    }
}
